### PR TITLE
Closes #190 and #191

### LINKED ISSUE #190: CMake Fortran compiler flags not set properly
The CMake Fortran compiler flags defined [here](https://github.com/mutationpp/Mutationpp/blob/c9a924fde43216e0777a918dc5042eaa18ae98b5/CMakeLists.txt#L153-163) are not set properly because the calls to `set` are missing the `CACHE` and `FORCE` attributes.  The `CMAKE_<LANG>_FLAGS*` are defined by CMake and are cached, so we need the `CACHE` flag when setting them explicitly to override the defaults. We also need the `FORCE` attribute to override the CMake default values for these variables.

### LINKED ISSUE #191: Fortran interface install destination undefined when using CMake command line
Using CMake via the command-line to build the Fortran wrapper, e.g., `cmake -DBUILD_FORTRAN_WRAPPER="TRUE"` results in an error. This is because `INSTALL_LIB_DIR` and `INSTALL_INCLUDE_DIR` variables are undefined when the Fortran interface CMakeLists is parsed, via the `add_subdirectory` call [here](https://github.com/mutationpp/Mutationpp/blob/c9a924fde43216e0777a918dc5042eaa18ae98b5/CMakeLists.txt#L47). The `INSTALL_LIB_DIR` and `INSTALL_INCLUDE_DIR` variables are not defined until later in the main CMakeLists.txt, [here](https://github.com/mutationpp/Mutationpp/blob/c9a924fde43216e0777a918dc5042eaa18ae98b5/CMakeLists.txt#L110-115). 

This can be resolved by adding the Fortran subdirectory after these variables are defined.

### PR DESCRIPTION
**Pull request by mgoodson-cvd**
Closes #190. Fixes an issue with Fortran compiler flags not being set properly.

Closes #191. Fixes an issue with the Fortran interface install directories not being set properly.

### CODE REVIEW

**Comment by rubendibattista** [comment]
Hello @mgoodson-cvd, thanks! I don't see the change to the install directories (I thought I had fixed it). Also no need to cache compiler flags.

**Comment by rubendibattista** [comment]
Also if you can it might be cool to also enable the build of the Fortran wrapper in the CI checks such as to catch issues related to Fortran in the CI

**Comment by mgoodson-cvd** [comment]
I'm not changing the behavior of the install directories; I'm changing when CMake steps into the Fortran subdirectory because where it is now is too early; the Fortran CMake [tries to use `INSTALL_LIB_DIR` and `INSTALL_INCLUDE_DIR`](https://github.com/mutationpp/Mutationpp/blob/c9a924fde43216e0777a918dc5042eaa18ae98b5/interface/fortran/CMakeLists.txt#L43-45) but they haven't been defined yet in the main CMakeLists.txt. I just moved the call to `add_subdirectory(interface/fortran)` to *after* these variables have been set.

File: CMakeLists.txt
Comment (rubendibattista) [inline CMakeLists.txt:162]: Ok, thanks for your contribution. Keeping the discussion here: 
> Also, it's worth pointing out that the CMAKE_CXX_FLAGS_Profile are force cached in that same file and they function as expected. I duplicated this behavior in my commit.

Ok, I think we shouldn't do that. Could you remove it also in the other places where they are cached?

File: CMakeLists.txt
Comment (rubendibattista) [inline CMakeLists.txt:173]: Ok, thanks for your contribution. Keeping the discussion here:

    Also, it's worth pointing out that the CMAKE_CXX_FLAGS_Profile are force cached in that same file and they function as expected. I duplicated this behavior in my commit.

Ok, I think we shouldn't do that. Could you remove it also in the other places where they are cached?

**Review by rubendibattista — CHANGES_REQUESTED** [review]
Also, if you can, enable the build of the Fortran wrapper in the Github actions such that we catch errors during CI testing.

File: CMakeLists.txt
Comment (mgoodson-cvd) [reply CMakeLists.txt:162]: I think you do want/need them cached. The `CMake_<LANG>_Flags*` already exist in the cache because they are defined by CMake when you enable the language `<LANG>`. You want to override the existing flags with the new compiler options, hence the `FORCE`; otherwise the defaults remain. And because these flags are already cached variables, you want the `CACHE`. I don't think you need the `CACHE` if you're strictly using the command line version of CMake; but if you're using the GUI, these variables *must* have the `CACHE` attribute. I do use the GUI, hence how I encountered this issue.

Comment (rubendibattista) [reply CMakeLists.txt:162]: Ok, I just tried (moving the `add_subdirectory`) as you did because I reproduced the install directory bug), and this is what I get without CACHE. 

`cmake -DBUILD_FORTRAN_WRAPPER=1 ..`

```
cd /home/rdb/git/Mutationpp/build/interface/fortran && /usr/sbin/gfortran -Dmutation___fortran_EXPORTS -I/home/rdb/git/Mutationpp/src/general -I/home/rdb/git/Mutationpp/src/kinetics -I/home/rdb/git/Mutationpp/src/numerics -I/home/rdb/git/Mutationpp/src/thermo -I/home/rdb/git/Mutationpp/src/transfer -I/home/rdb/git/Mutationpp/src/transport -I/home/rdb/git/Mutationpp/src/utilities -I/home/rdb/git/Mutationpp/src/gsi -I/usr/include/eigen3 -fdefault-real-8 -O3 -J. -fPIC -c /home/rdb/git/Mutationpp/interface/fortran/cwrapper_interface.f90 -o CMakeFiles/mutation++_fortran.dir/cwrapper_interface.f90.o
```
cmake -DBUILD_FORTRAN_WRAPPER=1 -DCMAKE_BUILD_TYPE=Debug ..

```
cd /home/rdb/git/Mutationpp/build/interface/fortran && /usr/sbin/gfortran -Dmutation___fortran_EXPORTS -I/home/rdb/git/Mutationpp/src/general -I/home/rdb/git/Mutationpp/src/kinetics -I/home/rdb/git/Mutationpp/src/numerics -I/home/rdb/git/Mutationpp/src/thermo -I/home/rdb/git/Mutationpp/src/transfer -I/home/rdb/git/Mutationpp/src/transport -I/home/rdb/git/Mutationpp/src/utilities -I/home/rdb/git/Mutationpp/src/gsi -I/usr/include/eigen3 -fdefault-real-8 -g -J. -fPIC -c /home/rdb/git/Mutationpp/interface/fortran/cwrapper_interface.f90 -o CMakeFiles/mutation++_fortran.dir/cwrapper_interface.f90.o
cd /home/rdb/git/Mutationpp/build/interface/fortran && /usr/sbin/c++ -Dmutation___fortran_EXPORTS -I/home/rdb/git/Mutationpp/src/general -I/home/rdb/git/Mutationpp/src/kinetics -I/home/rdb/git/Mutationpp/src/numerics -I/home/rdb/git/Mutationpp/src/thermo -I/home/rdb/git/Mutationpp/src/transfer -I/home/rdb/git/Mutationpp/src/transport -I/home/rdb/git/Mutationpp/src/utilities -I/home/rdb/git/Mutationpp/src/gsi -isystem /usr/include/eigen3 -g -g -fPIC -std=gnu++14 -MD -MT interface/fortran/CMakeFiles/mutation++_fortran.dir/cwrapper.cpp.o -MF CMakeFiles/mutation++_fortran.dir/cwrapper.cpp.o.d -o CMakeFiles/mutation++_fortran.dir/cwrapper.cpp.o -c /home/rdb/git/Mutationpp/interface/fortran/cwrapper.cpp
```

So for me flags are being passed correctly without CACHE. Can you please try from a clean `build` directory? Because I rarely saw compiler flags force cached. Also I don't see why things should be different between CLI / GUI. 

PS: The `-g` flag is passed twice.

Comment (mgoodson-cvd) [reply CMakeLists.txt:162]: Okay, this is somewhat confusing... I am able to repeat what you show above when I drop my second commit and use the CMake CLI -- the Fortran flags are set properly. When I use the GUI, the values displayed in the GUI for the compiler flags are incorrect; they are the default (cached) variables, because the GUI relies on the cache.

image

Note though that the `CMAKE_CXX_FLAGS_PROFILE` are set to the updated values from [here](https://github.com/mutationpp/Mutationpp/blob/c9a924fde43216e0777a918dc5042eaa18ae98b5/CMakeLists.txt#L66), because they are cached.

 However, when I actually look at the flags the compiler is using in `interface/fortran/CMakeFiles/mutation++_fortran.dir/flags.make`, **_they are set properly_**.

image

After reading more about the [GUI and the cache](https://cmake.org/cmake/help/latest/guide/user-interaction/index.html#id13), it seems the GUI is best viewed as an editor for the cache.
I'll let you decide how you want to proceed. I don't think there is anything wrong with caching the variables, especially since cached versions of them exist, and I think it would be less confusing to anyone using the GUI.

Comment (rubendibattista) [reply CMakeLists.txt:162]: Yeah, that's why I hate GUIs. They very often make things more confusing. :)

Anyway, if you tell me that force caching those flags shows them correctly in the GUI as you would expect, I can be good with that! In that case be sure that all the flags are correctly cached both for Fortran and C++

Comment (rubendibattista) [reply CMakeLists.txt:162]: Anyway, I think you need to run the configure step twice to get the right values in the GUI. It's how it works. The first time it shows the default because the local cache did not get generated.

Comment (rubendibattista) [reply CMakeLists.txt:162]: If also force caching does not show correct values in the GUI, then let's not cache them. :)

**Comment by rubendibattista** [comment]
I merged the fix on the install directories with #193. If you manage to check if without caching anything, the flags are correctly shown in the GUI after the second run of cmake, then I'd prefer to not cache them. If not, we will.

In both cases feel free to open another PR to address this :)

**Comment by mgoodson-cvd** [comment]
@rubendibattista I just pulled master after your merge of #193 and tried configuring using the CMake GUI (`ccmake`). With the compiler flags not cached, they are not shown correctly in the GUI; and now the C++ flags are also not shown correctly due to the modifications made in ab639ae0.

image

As before, this is just cosmetic; the *actual* compiler flags are set properly. The confusing part would be if the user tried to modify these flags via the GUI, they wouldn't actually be modified (I just tested this). And yes, you do have to "configure" multiple times in the GUI, otherwise the "generate" option never appears and you can't build -- you can see I did that properly in the screenshot.

I will open a new PR to cache the Fortran compiler flags and re-cache the C++ compiler flags.

**Pull request closed without merging**
